hpcrawler.py

```diff
@@ -0,0 +1,364 @@
+                            * The dir is named by the country name, string: <country>
+                            * The gpx files are named by the track's idx, int >= 0: <j>
+                            * The json file is named <progress>: maps [track_gpx_filename, track_length,
+                                track_difficulty] to track idx <j>, int >=0.
+     (2) dir <tracks_dir_path> : contains json files. These files are named by the length_tag of the tracks
+                                they hold: int >= 0 denoted as <l>.
+                                each json files holds a dict mapping the track name: <country><j>
+                                to the track's representation.
+     (3) json file <seen_path> : Keeps trace of the mining progress: contains a dict that maps the progress
+                                to the country <country>:
+                                [idx_of_next_track_we_need_to_crawl, [<urls>]] if we're mid process,
+                                and 'Done' otherwise.
+    """
+
+    # static fields:
+    wait = 5
+    done_tag = 'DONE'
+    gpx_dir_path = 'hp\\gpx'
+    tracks_dir_path = 'hp\\tracks'
+    seen_path = 'hp\\seen.json'
+
+    def __init__(self, to_crawl):
+        """
+        crawls the trails data from "the hiking project"'s site.
+        it saves the data and the progress of crawling task under the "hp" directory.
+        :param to_crawl: python list of countries to crawl.
+                        the countries should start with a capital letter.
+        """
+
+        self._countries = to_crawl
+        self._country = None  # string
+        self._track_idx = str(0)  # always string
+        self._path = None  # the location of collected data: gpx and progress of collecting (not final data)
+        self._url = None  # python list of strings
+        self._driver = None  # firefox(!) driver
+        self._driver_status = False
+
+    def __del__(self):
+        """
+        closes the driver in case an exception that hasn't been treated has occurred
+        """
+        if self._driver_status is True:
+            self._driver.quit()
+
+# writing to, and reading from, dicts in json files #
+    @staticmethod
+    def _load_dict(path):
+        """
+        :param path: a json file path
+        :return: dict obj, that was stored at <path>
+        """
+        dictionary = {}
+        if os.path.exists(path):
+            with open(path, "r") as f:
+                file = f.read()
+            dictionary = json.loads(file)
+        return dictionary
+
+    @staticmethod
+    def _save_dict(dictionary, path):
+        """
+        write back the dictionary <dict> to the json file at <path>
+        :param dictionary: a dictionary
+        :param path: a json file path
+        """
+        with open(path, 'w') as f:
+            json.dump(dictionary, f)
+
+    @staticmethod
+    def load_seen():
+        return HpCrawler._load_dict(HpCrawler.seen_path)
```

Comment:
this function is unnecessary, could rename HpCrawler._load_dict and have it receiver a default paramety of HpCrawler.seen_path
(I'm not sure if this is a must change, one could argue this is a style choice)